clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2838,11 +2835,14 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
   // kernel parameters from __init method parameters. We will use __init method
   // and kernel parameters which we build here to initialize special objects in
   // the kernel body.
-  bool handleSpecialType(FieldDecl *FD, QualType FieldTy) {
-    const auto *RecordDecl = FieldTy->getAsCXXRecordDecl();
+  // ParentDecl parameterizes whether we are in a free function kernel or a
+  // lambda kernel by taking the value ParmVarDecl or FieldDecl respectively.
+  template <typename ParentDecl>
+  bool handleSpecialType(ParentDecl *decl, QualType Ty) {
+    const auto *RecordDecl = Ty->getAsCXXRecordDecl();
```

Comment:
I don't know how did that compile, but there is RecordDecl type in clang, so I would recommend to rename this variable.
```suggestion
    const auto *RD = Ty->getAsCXXRecordDecl();
```